### PR TITLE
Inception metric

### PR DESCRIPTION
**Pull request by Devin-Taylor**
* implement inception model fine-tuning for individual experiments
* implement inception score evaluation metric

### CODE REVIEW

File: shenanigan/metrics/core.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    path = f"results/{experiment_name}/inception/model"
+    model = load_model(path)
+    processed = np.asarray((images + 1) * 255. / 2, np.uint8)
+    processed = preprocess_input(processed)
+    yhat = model.predict(processed, batch_size=1)
+    yhat = softmax(yhat, axis=1)
+    scores = list()
+    n_part = int(np.floor(images.shape[0] / n_split))
+    for i in range(n_split):
```
Comment (alecokas) [inline shenanigan/metrics/core.py:20]: I don't really understand why we have to split this up?

File: shenanigan/models/inception/model.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import tensorflow as tf
+from tensorflow.keras.applications import InceptionV3
+from tensorflow.keras.layers import Dense
+from typing import Tuple
```
Comment (alecokas) [inline shenanigan/models/inception/model.py:4]: Nice

File: shenanigan/models/inception/run.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    parsed_features = tf.io.parse_single_example(proto, f)
+    img = tf.io.decode_image(parsed_features['image_large'], dtype=tf.float32)
+    label = tf.one_hot(parsed_features['label'], depth=classes)
+    return img, label
+
+def load_dataset(input_path, batch_size, shuffle_buffer, classes):
+    dataset = tf.data.TFRecordDataset(input_path)
+    dataset = dataset.map(lambda x: _parse_function(x, classes), num_parallel_calls=16)
+    dataset = dataset.shuffle(shuffle_buffer).batch(batch_size).prefetch(1)
```
Comment (alecokas) [inline shenanigan/models/inception/run.py:28]: Only pre-fetch 1? Might this be faster if that is increased?

File: shenanigan/models/stackgan/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        _, _, large_images, _, _ = tensors_from_sample(
+            sample, batch_size, dataloader.dataset_object.text_embedding_dim, num_samples, augment
+        )
+        all_images += tf.squeeze(large_images).numpy().tolist()
+        # TODO move below code out of for loop once we can generate images from stage 2
+        # NOTE this uses a lot of memory so execute carefully
+        all_images = np.array(all_images)
+        is_avg, is_std = inception_score(experiment_name, all_images)
+        print("Inception score: ", is_avg, "+/-", is_std)
+        break
```
Comment (alecokas) [inline shenanigan/models/stackgan/evaluate.py:22]: Do you maybe mean to return something rather than just breaking? It might be good to save that score to file or something

File: shenanigan/models/stackgan/run.py [hunk 5]
```diff
@@ -103,6 +104,16 @@ def run(train_loader, val_loader, small_image_dims, results_dir, settings, stage
...
 
+    if evaluate:
+        eval_fxn(
+            model=model,
+            dataloader=val_loader,
+            experiment_name=experiment_name,
+            num_samples=settings['stage2']['num_samples'],
+            augment=True,
+            noise_size=settings['stage1']['noise_size']
+        )
```
Comment (alecokas) [inline shenanigan/models/stackgan/run.py:115]: This looks like it is meant to be used for evaluating the full model. Perhaps we need to block this if a stage 1 model is selected in the CLI.

File: shenanigan/metrics/core.py (hunk 1 above)
Comment (Devin-Taylor) [reply shenanigan/metrics/core.py:20]: original algorithm does, think it is to get a mean and std dev out of it. cross-validation-ish

File: shenanigan/models/inception/run.py (hunk 3 above)
Comment (Devin-Taylor) [reply shenanigan/models/inception/run.py:28]: Yeah I think it will, except it crashed my machine haha accidentally had batch size set to 640 so that was the cuprite. I saw somewhere recommend prefetching "batch size" but think that will kill the RAM. Will increase it to like 2-5 for now.

File: shenanigan/models/stackgan/evaluate.py (hunk 4 above)
Comment (Devin-Taylor) [reply shenanigan/models/stackgan/evaluate.py:22]: read the comments a few lines up, this is just for testing purposes. I actually need to run this with a stage 2 model, which I did not have available. This part is still WIP

File: shenanigan/models/stackgan/run.py (hunk 5 above)
Comment (Devin-Taylor) [reply shenanigan/models/stackgan/run.py:115]: good point, will make the change

File: shenanigan/metrics/InceptionScore.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import os
```
Comment (alecokas) [inline shenanigan/metrics/InceptionScore.py:1]: The file should be named with snake case

File: shenanigan/models/stackgan/run.py [hunk 7]
```diff
@@ -105,6 +106,21 @@ def run(train_loader, val_loader, small_image_dims, results_dir, settings, stage
         plotter = LogPlotter(results_dir)
         plotter.learning_curve()
 
+    if stage == 1 and evaluate:
+        raise NotImplementedError('Evaluation for stage 1 is not implemented')
```
Comment (alecokas) [inline shenanigan/models/stackgan/run.py:110]: Maybe do something like this for stage 2 as well if we can't evaluate stage 2

Comment (Devin-Taylor) [reply shenanigan/models/stackgan/run.py:110]: one of the previous checks will catch stage 2 and retained so figured that was fine

File: shenanigan/metrics/InceptionScore.py (hunk 6 above)
Comment (Devin-Taylor) [reply shenanigan/metrics/InceptionScore.py]: @alecokas fixed
